[CSS] Implement CSSOM insertRule() on StyleRule

mdubet opened the pull request:
#### 90d00dc15a7b02dda4a6088e22cbdc363b66e567
<pre>
[CSS] Implement CSSOM insertRule() on StyleRule
<a href="https://bugs.webkit.org/show_bug.cgi?id=254601">https://bugs.webkit.org/show_bug.cgi?id=254601</a>
rdar://107322119

Reviewed by Antti Koivisto.

This patch handles the transformation of old-school StyleRule object (without a children vector)
to a StyleRuleWithNesting object when necessary.

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalid-inner-rules-expected.txt:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
(WebCore::CSSGroupingRule::prepareChildStyleRuleForNesting):
* Source/WebCore/css/CSSGroupingRule.h:
* Source/WebCore/css/CSSRule.cpp:
(WebCore::CSSRule::hasStyleRuleAncestor const):
(WebCore::CSSRule::prepareChildStyleRuleForNesting):
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::cssTextForDeclsAndRules const):
(WebCore::CSSStyleRule::insertRule):
(WebCore::CSSStyleRule::deleteRule):
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::prepareChildStyleRuleForNesting):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::m_originalSelectorList):
(WebCore::StyleRuleWithNesting::StyleRuleWithNesting):
(WebCore::StyleRuleWithNesting::create):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isStyleRule const):
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseRule):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::CSSParserImpl):
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::isNestedContext):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeNestedSelectorList):

Canonical link: <a href="https://commits.webkit.org/262394@main">https://commits.webkit.org/262394@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ccfeda58e0829671bef12d783629a6a00d2bfb9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1342 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2358 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1327 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1433 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1441 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1393 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1354 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1256 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1239 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2092 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1269 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1224 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1204 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1227 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1259 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2392 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1294 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1166 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1208 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1231 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/355 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1311 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

shivamidow commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -82,6 +83,11 @@ ExceptionOr<unsigned> CSSGroupingRule::insertRule(const String& ruleString, unsi
     }
+
+    // Nesting inside style rule only accept style rule or group rule
+    if (hasStyleRuleAscendant() && (!newRule->isStyleRuleWithNesting() && !newRule->isStyleRule() && !newRule->isGroupRule()))
```
/* Unofficial Review */
Nit: The inner parenthesis seems unnecessary.

shivamidow commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -161,6 +167,21 @@ void CSSGroupingRule::cssTextForDeclsAndRules(StringBuilder& decls, StringBuilde
+    for (auto& rule : rules) {
+        if (rule == initialRule) {
+            rule = StyleRuleWithNesting::create(WTFMove(*styleRule));
+            return rule;
```
/* Unofficial Review */
Why do we assign a return value to a local variable?
```
return StyleRuleWithNesting::create(WTFMove(*styleRule));
```

mdubet replied:
`rule` is a reference to an object in a vector attribute (not to a local object), the goal of this method is specifically to modify it (and return the copy for the caller also).

shivamidow replied:
I see. Thanks for the clarification.

shivamidow commented on Source/WebCore/css/CSSStyleSheet.cpp:
```diff
@@ -176,6 +176,21 @@ Node* CSSStyleSheet::ownerNode() const
+    for (auto& rule : rules) {
+        if (rule == initialRule) {
+            rule = StyleRuleWithNesting::create(WTFMove(*styleRule));
+            return rule;
```
/* Unofficial Review */
Ditto.

anttijk commented on Source/WebCore/css/CSSRule.cpp:
```diff
@@ -58,4 +58,21 @@ const CSSParserContext& CSSRule::parserContext() const
     return styleSheet ? styleSheet->contents().parserContext() : strictCSSParserContext();
 }
 
+bool CSSRule::hasStyleRuleAscendant() const
```
"ancestor" is a more common word for this.

anttijk commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -161,6 +167,21 @@ void CSSGroupingRule::cssTextForDeclsAndRules(StringBuilder& decls, StringBuilde
     }
 }
 
+RefPtr<StyleRuleBase> CSSGroupingRule::transformStyleRuleToNesting(StyleRuleBase* initialRule)
```
This should take a reference since the parameter can never be null.

anttijk commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -161,6 +167,21 @@ void CSSGroupingRule::cssTextForDeclsAndRules(StringBuilder& decls, StringBuilde
+RefPtr<StyleRuleBase> CSSGroupingRule::transformStyleRuleToNesting(StyleRuleBase* initialRule)
+{
+    ASSERT(initialRule->isStyleRule());
+    auto styleRule = downcast<StyleRule>(initialRule);
```
The passed in parameter type should be StyleRule so you don't need to cast. That's the only type for which calling the function makes sense.

anttijk commented on Source/WebCore/css/CSSRule.h:
```diff
@@ -47,6 +47,8 @@ class CSSRule : public RefCounted<CSSRule> {
     CSSStyleSheet* parentStyleSheet() const;
     CSSRule* parentRule() const { return m_parentIsRule ? m_parentRule : nullptr; }
+    bool hasStyleRuleAscendant() const;
+    virtual RefPtr<StyleRuleBase> transformStyleRuleToNesting(StyleRuleBase*);
```
The full function signature should be like

`RefPtr<StyleRuleWithNesting> prepareChildStyleRuleForNesting(StyleRule&);`

The will substantially reduce need fo casting and documents the inputs and outputs.

anttijk commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -277,6 +282,7 @@ class StyleRulePage final : public StyleRuleBase {
 class StyleRuleGroup : public StyleRuleBase {
 public:
     const Vector<RefPtr<StyleRuleBase>>& childRules() const;
+    Vector<RefPtr<StyleRuleBase>>& childRules() { return m_childRules; }
```
It would be better to keep mutable vector private and expose it via `friend` to the function/class that needs this only. Making mutable fields easily available increases chance of future mistakes.

anttijk commented on Source/WebCore/css/StyleSheetContents.h:
```diff
@@ -107,6 +107,7 @@ class StyleSheetContents final : public RefCounted<StyleSheetContents>, public C
     const Vector<RefPtr<StyleRuleImport>>& importRules() const { return m_importRules; }
     const Vector<RefPtr<StyleRuleNamespace>>& namespaceRules() const { return m_namespaceRules; }
     const Vector<RefPtr<StyleRuleBase>>& childRules() const { return m_childRules; }
+    Vector<RefPtr<StyleRuleBase>>& childRules() { return m_childRules; }
```
same here

mdubet committed:
[CSS] Implement CSSOM insertRule() on StyleRule
https://bugs.webkit.org/show_bug.cgi?id=254601
rdar://107322119

Reviewed by Antti Koivisto.

This patch handles the transformation of old-school StyleRule object (without a children vector)
to a StyleRuleWithNesting object when necessary.

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalid-inner-rules-expected.txt:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
(WebCore::CSSGroupingRule::prepareChildStyleRuleForNesting):
* Source/WebCore/css/CSSGroupingRule.h:
* Source/WebCore/css/CSSRule.cpp:
(WebCore::CSSRule::hasStyleRuleAncestor const):
(WebCore::CSSRule::prepareChildStyleRuleForNesting):
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::cssTextForDeclsAndRules const):
(WebCore::CSSStyleRule::insertRule):
(WebCore::CSSStyleRule::deleteRule):
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::prepareChildStyleRuleForNesting):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::m_originalSelectorList):
(WebCore::StyleRuleWithNesting::StyleRuleWithNesting):
(WebCore::StyleRuleWithNesting::create):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isStyleRule const):
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseRule):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::CSSParserImpl):
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::isNestedContext):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeNestedSelectorList):

Canonical link: https://commits.webkit.org/262394@main
